### ISSUE TITLE
 minetest.register_timed_event

### ISSUE THREAD

**Issue by o-jasper** [minetest/minetest#4070]
Priority queue.. actualy using bins probably more efficient. (psuedocode.. i think)

    -- Note: suppose some locals here could go on the minetest object instead..
    local time, time_courseness = 0  
    local function bin_index(time)
        return math.floor(time/time_courseness)
    end
    local timed_event_bins, last_i = {}, 0
    minetest.register_globalstep(function(dt)
        time = time + dt
        local i = bin_index(time)
        if i ~= last_i  then
            last_i = i
            local list = timed_event_bins[i] or {}
            while true do
                local fun = table.remove(list)
                if not fun then
                    timed_event_bins[i] = nil  -- Done with the list.
                    return
                end
                fun()
            end
        end
    end)

    function minetest.register_timed_event(to_time, fun)
        assert(type(fun) == "function") -- Though a table with metatable would be callable too.
        local i = bin_index(to_time)
        if i <= last_i then return end  -- Already in the past.
        local list = timed_event_bins[i]
        if list then
            table.insert(list, fun)
        else  -- New list.
            timed_event_bins[i] = {fun}
        end
    end

Basically one `minetest.register_globalstep` can do a lot of timed events. Randomly timed events will usually be more efficient by random-generating the time once and inserting it here. "random decay" probability with a distribution as P(t)=exp(-t/T)/T has p=P(t'≤t)=1-exp(-t/T)
generating p∈[0,1] uniformly producing t=-T⋅log(1-p) has the distribution of P(t); the density is 1/(dt/dp) = (1-p)/T = (1-(1-exp(-t/T)))/T = exp(-t/T)/T = P(t), other random distributions can be done too.

Could be used instead of(and more flexible then) [on_timer](http://dev.minetest.net/minetest.register_node#on_timer).

**Comment by sofar** [comment]
I can see value in this, similar ideas like this are in the linux kernel for instance.

Why not however change the existing timer code to allow for ```timerslack``` to be part of the current timer code, though? (see e.g. https://lwn.net/Articles/369549/ )

The above code doesn't pass parameters to fun(), time_courseness is hardcoded to 0? Don't you want to calculate it with a formula for each ```time``` instead?

You need to walk the list backwards to prevent iterating issues. Using table.insert() and table.remove() is potentially slow.

This needs a lot more explanation and detail, I think, as well. Provide use cases?

**Comment by o-jasper** [comment]
I noticed an issue; `minetest.register_node` is in the definition, if an object is "frozen out of the current simulation", it is just map data. *Here*, the function is attached on-the-fly the data is lua-data, so cannot readily be stored. Perhaps force referring to a definition somehow, and have `fun` receive data about whatever object it was about. However #3835 is resolved might include what about that is "permanent".

Nevertheless i notice in `src/nodetimer.cpp` that it works by **incrementing all the timers and checking every timestep**. So think the above method could still be faster.

`time_courseness` is `nil`, oops. Also *if* the timestep is larger than the courseness, then you must make sure you get them all. Really though, might aswel make all timesteps equal and a bin for each?

Trying to be terse, so for usecase, i.e. trees already grow after a time presumably using `on_timer`, it is not really new functionality, other than basically asking to untie it from the node. `timerslack`, is interesting link, but probably not applicable. We're not really talking about skipping timesteps here?

Possibly intermediate conclusion might be to 1) change the algo in/around `src/nodetimer.cpp` 2) make `NodeTimer` less node-specific, and allow anything to have a, or multiple timers.

**Comment by o-jasper** [comment]
Note: might counter that a particle takes *a lot* more computation than incrementing one thing. But it is still useful to make it more efficient because it allows orders of magnitudes more timers.

One thing every 10 surface blocks. continuously simulate 1000×1000 of surface around the origin, thats 0.1M things. Once on average happens 10mins, then a mere 170 events/second. In comparison upticking each with a 10ms timestep you need to iterate over 10M/second. (Memory use is high; 10GB assuming 100nodes high and 100bytes/each.

Essentially, it pre-calculates all the time instances together. Can also take all the nodes of the same type together. Though afaics for anything other than exponential decay you'd need age-bins, and you'd need to do the math for the next-time-one-of-N does something.(Likely only worth it if there are *many* of them)

Note that also, to simulate "the land", if you also want to have mesecon, it'd help a lot if the mese is disabled when no-one is around.(not part of "the land")

*Anyway* back to the matter at hand, the above can probably be done with lua, given timers, so not particularly engine-y. So `NodeTimerList` already owns `remove`/`set` it could do bins afaics. However, each `MapBlock` has one, whereas it is better if the entire simulation has one. It is also serialized-per-block. Probably want to keep that.. hmm, might mean ferretting around searching for the associated timers in the global simulation serializing things. Also auch, soo much code here... I should have realized this better before starting this..

**Comment by o-jasper** [comment]
#3895 gets the efficiency, i think.

Still attached to nodes, imo better if some concept of a timer, attached equally to other objects? (players/entities/nodes/"the world")
